contracts/modules/capital/Pool.sol

```diff
@@ -166,44 +196,93 @@ contract Pool is IPool, MasterAware, ReentrancyGuard {
+    uint payoutAssetsCount = payoutAssets.length;
+    for (uint i = 0; i < payoutAssetsCount; i++) {
+      require(assetAddress != payoutAssets[i].assetAddress, "Pool: Asset exists");
+    }
+
+    // Check whether the new asset already exists as an investment asset
+    uint investmentAssetsCount = investmentAssets.length;
+    for (uint i = 0; i < investmentAssetsCount; i++) {
+      require(assetAddress != investmentAssets[i].assetAddress, "Pool: Asset exists");
     }
 
-    assets.push(Asset(assetAddress, decimals, false));
+    // Add the new asset to its corresponding array
+    if (isPayoutAsset) {
+      payoutAssets.push(Asset(assetAddress, decimals));
+    } else {
+      investmentAssets.push(Asset(assetAddress, decimals));
+    }
+
+    // Set the swap details
     swapDetails[assetAddress] = SwapDetails(_min, _max, 0, _maxSlippageRatio);
   }
 
-  function deprecateAsset(uint8 assetId) external onlyGovernance {
-    require(assetId < assets.length, "Pool: Asset does not exist");
-    address assetAddress = assets[assetId].assetAddress;
-    delete swapDetails[assetAddress];
-    assets[assetId].deprecated = true;
+  /// Removes an asset which is no longer used.
+  ///
+  /// @dev Investment assets will be removed from the investmentAssets array. Payout assets
+  /// however cannot be removed entirely as they are referenced by their index in covers.
+  /// Instead, they are ignored by setting the bit corresponding to the asset's index in
+  /// ignoredPayoutAssets to 1. Ignored payout assets are skipped when calculating the pool value
+  /// in ETH which saves a slot read for each asset removed. However this does not prevent cover
+  /// sales in that particular payout asset and it is required to set coverAssetsFallback and
+  /// coverAssets on each product beforehand (See: Cover.sol). When an asset is removed, the
+  /// corresponding swapDetails are also removed and it is assumed that the balance is 0. To allow
+  /// removing assets which might revert on balance calls (such as a malicious ), there are no
+  /// balance checks in this function.
+  ///
+  /// @param assetId        The index of the asset that needs to be removed.
+  /// @param isPayoutAsset  True if the asset is used for payouts or false if it's just an
+  ///                       investment asset.
+  ///
+  function removeAsset(uint assetId, bool isPayoutAsset) external onlyGovernance {
+    if (isPayoutAsset) {
+      require(assetId < payoutAssets.length, "Pool: Payout asset does not exist");
+      require(ignoredPayoutAssetsBitmap & (1 << assetId) != 0, "Pool: Payout asset is ignored");
+
+      // Remove swap details
+      address assetAddress = payoutAssets[assetId].assetAddress;
+      delete swapDetails[assetAddress];
+
+      // Ignore asset which makes getPoolValueInEth skip it when the function loops through
+      // payments assets
+      ignoredPayoutAssetsBitmap |= SafeUintCast.toUint32(1 << assetId);
```

Comment:
This design has a terrible limitation that it will indeed prevent cover buys in the said asset but will also immediately exclude them from the pool value calculation.

Reply:
It's as "terrible" as the current mainnet implementation, where the same thing happens when you remove an asset. That's intended for both cover buys and pool value calculation. I don't see any actionables nor constructive arguments in this comment. Please clarify if necessary!

Reply:
Let's say tomorrow we want to prevent cover buys with DAI (for whatever reason). Currently we have over $2m DAI in pool, this will instantly impact the pool value. We need to be able to prevent cover buys while still using the asset in the pool value calculation. An additional issue would be that we still have to be able to be able to send payouts in that asset for 1 year + grace period. After this period we should be able to sell the asset from the pool and only then remove it.

It is indeed true that this issue is present in the current mainnet implementation, seems like we didn't notice until now, however that's not a reason to keep it that way.

Reply:
2 things here:

1. As stated in the natspec, the flag doesn't prevent cover buys. That's done separately in Cover.sol (coverAssetsFallback and coverAssets in the Product struct).
2. Added a check for balance requiring it to be 0 before removing the token. If the balanceOf function reverts it is considered 0.